File: src/main/scala/com/vc/medium/L1027.scala
package com.vc.medium

object L1027 {
  def longestArithSeqLength(A: Array[Int]): Int = {
    import scala.collection.mutable
    val dp = new Array[mutable.HashMap[Int, Int]](A.length)
    val n = A.length
    var res = 0
    (0 until n).foreach(i => {
      dp(i) = new mutable.HashMap[Int, Int]()
      (0 until i).foreach(j => {
        val d = A(j) - A(i)
        dp(i).put(d, dp(j).getOrElse(d, 1) + 1)
        res = Math.max(res, dp(i)(d))
      })
    })
    res
  }
}

File: src/main/java/com/vc/medium/L1027.java
package com.vc.medium;

import java.util.*;

class L1027 {
    public int longestArithSeqLength(int[] A) {
        int n = A.length, max = 0;
        HashMap<Integer, Integer>[] arr = new HashMap[n];

        arr[0] = new HashMap<Integer, Integer>();
        for(int i = 1; i < n; i++) {
            arr[i] = new HashMap<Integer, Integer>();
            int current = A[i];
            for(int j = 0; j < i; j++) {
                int diff = current - A[j];
                arr[i].put(diff, arr[j].getOrDefault(diff, 0) + 1);
                max = Math.max(max, arr[i].get(diff));
            }
        }
        return max + 1;
    }
}
